### PR TITLE
Ensure that USE statements are recognized and apply to table references without a qualifying schema in SQL and pyspark

### LINKED ISSUE #1215: [FEATURE]: Detect and process `USE <database>` statements
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

USE statement is used to set the current database. After the current database is set, the unqualified database artifacts such as tables, functions and views that are referenced by SQLs are resolved from the current database. The default database name is ‘default’.

```
USE database_name
```

```
spark.sql("use something")
spark.table("anything").show()
```

the second statement would show records from `something.anything`, not `default.anything`.

See https://spark.apache.org/docs/3.0.0-preview/sql-ref-syntax-qry-select-usedb.html

The following test case has to produce three different table advice:

```
# Databricks notebook source
display(spark.table('people')) # we are looking at default.people table

# COMMAND ----------

# MAGIC %sql USE something

# COMMAND ----------

display(spark.table('persons')) # we are looking at something.persons table

# COMMAND ----------

spark.sql('USE whatever')

# COMMAND ----------

display(spark.table('kittens')) # we are looking at whatever.kittens table
```

### Proposed Solution

have linters/fixers pickup `USE database` statements if they occur in the code to resolve single-table-name lookups

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jimidle**
## Changes
The linters and migrators are now aware of USE statements in SQL and pyspark. Tables are no longer erroneously flagged or migrated because the parsers did not track use statements. 

### Linked issues

Resolves #1215 

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

**Comment by jimidle** [comment]
Please note that the current commits only change the SQL table migrator. Pyspark in the next commit.

File: src/databricks/labs/ucx/source_code/queries.py [hunk 1]
```diff
@@ -42,17 +49,22 @@ def lint(self, code: str) -> Iterable[Advice]:
                     end_col=1024,
                 )
 
-    @staticmethod
-    def _catalog(table):
+    def _catalog(self, table):
         if table.catalog:
             return table.catalog
-        return 'hive_metastore'
+        return self._use_schema if self._use_schema else 'hive_metastore'
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/queries.py:55]: This is incorrect, catalog is one level above the schema

File: src/databricks/labs/ucx/source_code/queries.py [hunk 2]
```diff
@@ -42,17 +49,22 @@ def lint(self, code: str) -> Iterable[Advice]:
...
 
     def apply(self, code: str) -> str:
         new_statements = []
         for statement in sqlglot.parse(code, read='databricks'):
             if not statement:
                 continue
+            if isinstance(statement, Use):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/queries.py:62]: Do we need this?

File: src/databricks/labs/ucx/source_code/queries.py [hunk 3]
```diff
@@ -65,14 +77,11 @@ def apply(self, code: str) -> str:
...
             new_statements.append(new_sql)
+        # TODO: Should the return preserve newlines?
         return '; '.join(new_statements)
 
-    @classmethod
-    def _dependent_tables(cls, statement: Expression):
+    def _dependent_tables(self, statement: Expression):
         dependencies = []
         for old_table in statement.find_all(Table):
-            catalog = cls._catalog(old_table)
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/queries.py:74]: Why did you remove this?

File: tests/unit/source_code/test_queries.py [hunk 4]
```diff
@@ -51,3 +52,26 @@ def test_fully_migrated_queries_match_no_db(migration_index):
     new_query = "SELECT * FROM brand.new.stuff LEFT JOIN some.certain.issues USING (x) WHERE state > 1 LIMIT 10"
 
     assert ftf.apply(old_query) == new_query
+
+
+def test_use_database_change(migration_index):
+    ftf = FromTable(migration_index, use_schema="old")
+    query = """
+    USE newcatalog;
```
Comment (nfx) [inline tests/unit/source_code/test_queries.py:60]: Schema is not catalog, change asap, otherwise it'll confuse everyone

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/queries.py (hunk 1 above)
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/queries.py:55]: Yes - schema and catalog seem to be used interchangeably sometimes in some documentation - I am used to catalog/db.schema.table. I was going to ask you about that.

File: src/databricks/labs/ucx/source_code/queries.py (hunk 2 above)
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/queries.py:62]: I think so. I need to know that the statement is a USE statement to switch the schema for subsequent statements and subsequent linters.

File: src/databricks/labs/ucx/source_code/base.py [hunk 5]
```diff
@@ -69,23 +69,46 @@ class Convention(Advice):
...
 
+# The default schema to use when the schema is not specified in a table reference
+# See: https://spark.apache.org/docs/3.0.0-preview/sql-ref-syntax-qry-select-usedb.html
+DEFAULT_SCHEMA = 'default'
+
+
 class Linter:
     @abstractmethod
-    def lint(self, code: str) -> Iterable[Advice]: ...
+    def lint(self, code: str, schema: str) -> Iterable[Advice]: ...
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/base.py:79]: i'm going to block the change to this interface.

File: src/databricks/labs/ucx/source_code/languages.py [hunk 6]
```diff
@@ -37,10 +37,10 @@ def fixer(self, language: Language, diagnostic_code: str) -> Fixer | None:
                 return fixer
         return None
 
-    def apply_fixes(self, language: Language, code: str) -> str:
+    def apply_fixes(self, language: Language, code: str, schema: str) -> tuple[str, str]:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/languages.py:40]: we need to introduce a new `CurrentSessionState` and inject it into two classes: `FromTable`, `SparkSql`. Do it from `Languages.__init__`. NOTE: _the instance of languages should be modified per notebook._

```
@dataclass
class CurrentSessionState:
   use_schema: str | None
```

File: src/databricks/labs/ucx/source_code/notebook_linter.py [hunk 7]
```diff
@@ -14,22 +14,35 @@ class NotebookLinter:
...
+    @property
+    def schema(self):
+        # This linter is a controller of a number of linters for each cell
+        # language in the notebook. It is not generally useful to know the schema
+        # of the last cell that was linted, except for testing purposes.
+        return self._schema
 
     @classmethod
     def from_source(cls, langs: Languages, source: str, default_language: Language) -> 'NotebookLinter':
         notebook = Notebook.parse("", source, default_language)
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook_linter.py:28]:
```suggestion
    def from_source(cls, index: MigrationIndex, source: str, default_language: Language) -> 'NotebookLinter':
        languages = Languages(index)
        notebook = Notebook.parse("", source, default_language)
```
this way we ensure that `CurrentSessionState` is regenerated very time we're linting new notebook

**Review by nfx — CHANGES_REQUESTED** [review]
don't change the `Linter` interface, introduce the `CurrentSessionState`.

**Comment by nfx** [comment]
Include this test case:
```
# Databricks notebook source
display(spark.table('people')) # we are looking at default.people table

# COMMAND ----------

# MAGIC %sql USE something

# COMMAND ----------

display(spark.table('persons')) # we are looking at something.persons table

# COMMAND ----------

spark.sql('USE whatever')

# COMMAND ----------

display(spark.table('kittens')) # we are looking at whatever.kittens table
```

File: src/databricks/labs/ucx/source_code/base.py [hunk 8]
```diff
@@ -69,23 +69,46 @@ class Convention(Advice):
...
+    def schema(self):
+        return None
 
 
 class Fixer:
     @abstractmethod
     def name(self) -> str: ...
 
     @abstractmethod
-    def apply(self, code: str) -> str: ...
```
Comment (FastLee) [inline src/databricks/labs/ucx/source_code/base.py:82]: What's the use of a property that returns null?

File: src/databricks/labs/ucx/source_code/dbfs.py [hunk 9]
```diff
@@ -65,14 +65,19 @@ class DBFSUsageLinter(Linter):
     def __init__(self):
         pass
 
+    @property
```
Comment (FastLee) [inline src/databricks/labs/ucx/source_code/dbfs.py:68]: I don't understand the use of this property

**Review by FastLee — CHANGES_REQUESTED** [review]
Add more tests.
Test for all three flavors of "use"
use database <schema name>
use schema <schema name>
use <schema name>

File: src/databricks/labs/ucx/source_code/base.py (hunk 8 above)
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/base.py:82]: This is no longer present with next push

File: src/databricks/labs/ucx/source_code/dbfs.py (hunk 9 above)
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/dbfs.py:68]: This is no longer present with the next push

File: src/databricks/labs/ucx/source_code/notebook_linter.py (hunk 7 above)
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/notebook_linter.py]: Done

Comment (jimidle) [reply src/databricks/labs/ucx/source_code/notebook_linter.py]: Linter interface is unchanged

File: src/databricks/labs/ucx/hive_metastore/view_migrate.py [hunk 10]
```diff
@@ -41,7 +41,7 @@ def _view_dependencies(self):
             yield TableView("hive_metastore", src_db, old_table.name)
 
     def sql_migrate_view(self, index: MigrationIndex) -> str:
-        from_table = FromTable(index, use_schema=self.src.database)
+        from_table = FromTable(index)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/view_migrate.py:44]: this breaks views migration. i wonder why `make test` didn't fail. 
```suggestion
        from_table = FromTable(index, CurrentSessionState(self.src.database))
```

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/dbfs.py [hunk 11]
```diff
@@ -65,14 +65,19 @@ class DBFSUsageLinter(Linter):
...
+    def schema(self):
```
[the remark this answers is not in the document]
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/dbfs.py:69]: we don't - I will remove

File: src/databricks/labs/ucx/source_code/dbfs.py [hunk 12]
```diff
@@ -65,14 +65,19 @@ class DBFSUsageLinter(Linter):
...
+
     @staticmethod
     def name() -> str:
         """
         Returns the name of the linter, for reporting etc
         """
         return 'dbfs-usage'
 
-    def lint(self, code: str) -> Iterable[Advice]:
+    def lint(self, code: str, _: str | None = None) -> Iterable[Advice]:
```
[the remark this answers is not in the document]
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/dbfs.py:80]: Sure. An oversight. I forget to put that bacl

File: src/databricks/labs/ucx/source_code/dbfs.py [hunk 13]
```diff
@@ -86,12 +91,17 @@ class FromDbfsFolder(Linter):
     def __init__(self):
         self._dbfs_prefixes = ["/dbfs/mnt", "dbfs:/", "/mnt/", "/dbfs/", "/"]
 
+    @property
+    def schema(self):
```
[the remark this answers is not in the document]
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/dbfs.py:95]: yup

File: src/databricks/labs/ucx/source_code/dbfs.py [hunk 14]
```diff
@@ -86,12 +91,17 @@ class FromDbfsFolder(Linter):
...
+        # This is a linter that does not need to know the schema
+        return None
+
     @staticmethod
     def name() -> str:
         return 'dbfs-query'
 
-    def lint(self, code: str) -> Iterable[Advice]:
-        for statement in sqlglot.parse(code, dialect='databricks'):
+    def lint(self, code: str, _: str | None = None) -> Iterable[Advice]:
```
[the remark this answers is not in the document]
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/dbfs.py:103]: yup

File: src/databricks/labs/ucx/source_code/notebook_linter.py [hunk 15]
```diff
@@ -14,14 +15,16 @@ class NotebookLinter:
...
     @classmethod
-    def from_source(cls, langs: Languages, source: str, default_language: Language) -> 'NotebookLinter':
+    def from_source(cls, index: MigrationIndex, source: str, default_language: Language) -> 'NotebookLinter':
+        langs = Languages(index)
         notebook = Notebook.parse("", source, default_language)
         assert notebook is not None
         return cls(langs, notebook)
 
-    def lint(self) -> Iterable[Advice]:
+    def lint(self, _: str) -> Iterable[Advice]:
```
[the remark this answers is not in the document]
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/notebook_linter.py:27]: yup

File: src/databricks/labs/ucx/source_code/python_linter.py [hunk 16]
```diff
@@ -180,6 +179,14 @@ def collect_appended_sys_paths(self):
 
 class PythonLinter(Linter):
 
+    def __init__(self):
+        self._schema = DEFAULT_SCHEMA
+
+    @property
```
[the remark this answers is not in the document]
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/python_linter.py:185]: missed it

File: src/databricks/labs/ucx/source_code/queries.py [hunk 17]
```diff
@@ -2,39 +2,71 @@
...
-    def __init__(self, index: MigrationIndex, *, use_schema: str | None = None):
-        self._index = index
-        self._use_schema = use_schema
+    """Linter and Fixer for table migrations in SQL queries.
+
+    This class is responsible for identifying and fixing table migrations in
+    SQL queries.
+    """
+
+    def __init__(self, index: MigrationIndex, *, session_state: CurrentSessionState | None = None):
```
[the remark this answers is not in the document]
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/queries.py:19]: Yes - I was wavering on that. WIll do.

File: src/databricks/labs/ucx/source_code/queries.py [hunk 18]
```diff
@@ -2,39 +2,71 @@
...
+        for statement in sqlglot.parse(code, read='databricks'):
             if not statement:
                 continue
             for table in statement.find_all(Table):
-                catalog = self._catalog(table)
-                if catalog != 'hive_metastore':
+                if isinstance(statement, Use):
+                    # Sqlglot captures the database name in the Use statement as a Table, with
+                    # the schema  as the table name.
+                    self._session_state.schema = table.name if table.name else DEFAULT_SCHEMA
```
[the remark this answers is not in the document]
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/queries.py:53]: Ah. I was asked to add a check for USE without a database. So, that had to be checked for. Then I found sqlglot says that is impossible, which I had thought myself. That got left there.

File: src/databricks/labs/ucx/hive_metastore/view_migrate.py (hunk 10 above)
Comment (jimidle) [reply src/databricks/labs/ucx/hive_metastore/view_migrate.py:44]: Good spot

Comment (jimidle) [reply src/databricks/labs/ucx/hive_metastore/view_migrate.py:44]: I will check the tests and see if I can add one that would point this out

Comment (jimidle) [reply src/databricks/labs/ucx/hive_metastore/view_migrate.py:44]: There are no tests that would have caught that. Given the session state is now mandatory, perhaps there is no need for one?

File: src/databricks/labs/ucx/source_code/notebook_linter.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/source_code/notebook_linter.py:27]: Revert

File: src/databricks/labs/ucx/source_code/notebook_linter.py [hunk 19]
```diff
@@ -14,14 +15,16 @@ class NotebookLinter:
     def __init__(self, langs: Languages, notebook: Notebook):
         self._languages: Languages = langs
         self._notebook: Notebook = notebook
+        self._schema: str = DEFAULT_SCHEMA
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook_linter.py:18]: Why is it still here?

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/notebook_linter.py (hunk 19 above)
Comment (jimidle) [reply src/databricks/labs/ucx/source_code/notebook_linter.py]: Removed.

**Review by nfx — APPROVED** [review]
Lgtm
